ansible_bender/callback_plugins/snapshoter.py

```diff
@@ -69,17 +72,50 @@ def _snapshot(self, task_result):
 
     @staticmethod
     def get_task_content(task: Task):
+        sha512 = hashlib.sha512()
         serialized_data = task.get_ds()
         if not serialized_data:
             # ansible 2.8
             serialized_data = task.dump_attrs()
         if not serialized_data:
             logger.error("unable to obtain task content from ansible: caching will not work")
             return
-        c = json.dumps(serialized_data, sort_keys=True).encode("utf-8")
+        c = json.dumps(serialized_data, sort_keys=True)
+
         logger.debug("content = %s", c)
-        m = hashlib.sha512(c)
-        return m.hexdigest()
+        sha512.update(c.encode("utf-8"))
+
+        # If task is a file action, cache the src.
+        # Take the contents of the src (could be file or directory)
+        # and concatinate it with the task config (or serialized_data).
+        #
+        # Pseudocode: hash(serialized_data + src_file)
+        #
+        # Note: serialized_data was grabbed above.
+        try:
+            task_config = json.loads(c)[task.action]
+            src_regex = "(?<=src\=).*?(?=\ )"
+            src = re.search(src_regex, task_config).group(0)
+            src_path = task.get_search_path()[0] + "/files/" + src
+
+            if os.path.isdir(src_path):
+                # If it's a directory, hash the contents
+                # of everything in the directory.
+                # Then hash the return value.
+                dir_hash = dirhash(src_path, 'sha512')
+                sha512.update(dir_hash.encode("utf-8"))
+            elif os.path.isfile(src_path):
+                BUF_SIZE = 65536
+                with open(src_path, 'rb') as f:
+                    while True:
+                        data = f.read(BUF_SIZE)
+                        if not data:
+                            break
+                        sha512.update(data)
+        except:
```

Comment:
hmmm...

Reply:
That didn't end up doing what I hoped it would. The intent was that if the data extracted from the file was 'invalid', it would skip the caching and rebuild the layer. What really happens is that it reloads from cache no matter what, even if there were changes to the file. If any of that makes sense.

I'm checking the file stats (namely modification time) in the case of a singular file. Same as directories
```
            elif os.path.isfile(src_path):
                the_file = pathlib.Path(src_path)
                file_stats = str(the_file.stat())
                sha512.update(file_stats.encode("utf-8"))
```